src/snapred/backend/data/LocalDataService.py

```diff
@@ -323,496 +271,223 @@ def _constructReductionDataFilePath(self, runNumber: str, useLiteMode: bool, ver
+        Record must be set with correct version.
+        -- side effect: creates needed directories for save
         """
-        runNumber = record.runNumber
-        stateId, _ = self._generateStateId(runNumber)
-        previousVersion = self._getLatestNormalizationVersionNumber(stateId, record.useLiteMode)
-        if version is None:
-            version = max(record.version, previousVersion + 1)
-        recordPath: Path = self.getNormalizationRecordFilePath(runNumber, record.useLiteMode, version)
-        record.version = version
 
-        # There seems no need to write the _nested_ Normalization,
-        # because it's written to a separate file during 'writeNormalizationState'.
-        # However, if it is going to be _nested_, this marks it with the correct version.
-        # (For example, use pydantic Field(exclude=True) to _stop_ nesting it.)
-        record.calibration.version = version
-
-        normalizationPath: Path = self._constructNormalizationDataPath(runNumber, record.useLiteMode, version)
-        # check if directory exists for runId
-        if not normalizationPath.exists():
-            os.makedirs(normalizationPath)
-        # append to record and write to file
-        write_model_pretty(record, recordPath)
-        logger.info(f"wrote NormalizationRecord: version: {version}")
-        return record
+        indexer = self.normalizationIndexer(record.runNumber, record.useLiteMode)
+        # write the record to file
+        indexer.writeRecord(record)
+        # separately write the normalization state
+        indexer.writeParameters(record.calculationParameters)
 
-    def writeNormalizationWorkspaces(self, record: NormalizationRecord) -> NormalizationRecord:
+        logger.info(f"wrote NormalizationRecord: version: {record.version}")
+
+    def writeNormalizationWorkspaces(self, record: NormalizationRecord):
         """
         Writes the workspaces associated with a `NormalizationRecord` to disk:
+        Record must be set with correct version and workspace names finalized.
         -- assumes that `writeNormalizationRecord` has already been called, and that the version folder exists
         """
-        normalizationDataPath: Path = self._constructNormalizationDataPath(
-            record.runNumber, record.useLiteMode, record.version
-        )
+        indexer = self.normalizationIndexer(record.runNumber, record.useLiteMode)
+        normalizationDataPath: Path = indexer.versionPath(record.version)
+        if not normalizationDataPath.exists():
+            normalizationDataPath.mkdir(parents=True, exist_ok=True)
         for workspace in record.workspaceNames:
-            filename = workspace + "_" + wnvf.formatVersion(record.version)
             ws = mtd[workspace]
             if ws.isRaggedWorkspace():
-                filename = Path(filename + ".tar")
+                filename = Path(workspace + ".tar")
```

Comment:
Does this change mean that workspace names come with a version attached now?

Reply:
At this location in the `LocalDataService`, yes.  But they are set inside the `CalibrationService` with the same process.  I think Kort wants us to push it back at the time of workspace name creation, which isn't hard but maybe we can do later.